### PR TITLE
Open nodes

### PR DESCRIPTION
**Pull request by nebolax**
- [x] Change default ethereum nodes
- [x] Add migration to update nodes of existing users

### CODE REVIEW

File: rotkehlchen/data/nodes.json [hunk 1]
```diff
@@ -8,48 +8,32 @@
...
-        "weight": 0.15,
-        "owned": false,
-        "active": true,
-        "blockchain": "ETH"
-    },
-    {
-        "name": "cloudflare",
-        "endpoint": "https://cloudflare-eth.com/",
+        "endpoint": "https://nodes.mewapi.io/rpc/eth",
         "weight": 0.10,
```
Comment (yabirgb) [inline rotkehlchen/data/nodes.json:37]: What criteria did you use to rank the weight of the nodes? Did you perform any comparison?

File: rotkehlchen/data_migrations/migrations/migration_6.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    'https://api.mycryptoapi.com/eth',  # mycrypto
+    'https://mainnet-nethermind.blockscout.com/',  # blockscout
+    'https://mainnet.eth.cloud.ava.do/',  # avado
+    'https://web3.1inch.exchange',  # 1inch
+    'https://names.mewapi.io/rpc/eth',  # myetherwallet
+]
+
+
+def update_nodes_in_database(write_cursor: 'DBCursor', rotki: 'Rotkehlchen') -> None:
+    dir_path = Path(__file__).resolve().parent.parent.parent
```
Comment (yabirgb) [inline rotkehlchen/data_migrations/migrations/migration_6.py:28]: Can you add a docstring explaining what this migration does and the two possible cases

File: rotkehlchen/data_migrations/migrations/migration_6.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    nodes_in_db = set(write_cursor.execute(
+        'SELECT name, endpoint, owned, weight FROM web3_nodes WHERE blockchain=?',
+        (SupportedBlockchain.ETHEREUM.value,),
+    ))
+
+    if nodes_in_db == old_nodes:
+        # If nodes in the db are exactly same as old defaults, replace them with new defaults
+        write_cursor.execute(
```
Comment (yabirgb) [inline rotkehlchen/data_migrations/migrations/migration_6.py:46]: can you add a log entry to mark what branch is executed during the migration?

File: rotkehlchen/tests/data_migrations/test_migrations.py [hunk 4]
```diff
@@ -285,3 +285,97 @@ def test_migration_5(rotkehlchen_api_server):
...
+
+
+@pytest.mark.parametrize('data_migration_version', [None])
+@pytest.mark.parametrize('perform_migrations_at_unlock', [False])
+@pytest.mark.parametrize('perform_upgrades_at_unlock', [False])
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+@pytest.mark.parametrize('perform_nodes_insertion', [False])
+def test_migration_6_default_nodes(rotkehlchen_api_server):
+    """
+    Test that the sixs data migration works when the user has not customized the nodes.
```
Comment (yabirgb) [inline rotkehlchen/tests/data_migrations/test_migrations.py:297]:
```suggestion
    Test that the sixth data migration works when the user has not customized the nodes.
```

File: rotkehlchen/tests/data_migrations/test_migrations.py [hunk 5]
```diff
@@ -285,3 +285,97 @@ def test_migration_5(rotkehlchen_api_server):
...
+
+
+@pytest.mark.parametrize('data_migration_version', [None])
+@pytest.mark.parametrize('perform_migrations_at_unlock', [False])
+@pytest.mark.parametrize('perform_upgrades_at_unlock', [False])
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+@pytest.mark.parametrize('perform_nodes_insertion', [False])
+def test_migration_6_customized_nodes(rotkehlchen_api_server):
+    """
+    Test that the sixs data migration works when the user has some cusotmized the nodes.
```
Comment (yabirgb) [inline rotkehlchen/tests/data_migrations/test_migrations.py:344]:
```suggestion
    Test that the sixth data migration works when the user has cusotmized the rpc nodes.
```

File: rotkehlchen/tests/data_migrations/test_migrations.py [hunk 6]
```diff
@@ -285,3 +285,97 @@ def test_migration_5(rotkehlchen_api_server):
...
     assert Path(icons_path, '_ceth_0x6DEA81C8171D0bA574754EF6F8b412F2Ed88c54D_small.png').exists() is False  # noqa: E501
     assert Path(icons_path, '_ceth_0x7Fc66500c84A76Ad7e9c93437bFc5Ac33E2DDaE9_small.png').exists() is False  # noqa: E501
+
+
+@pytest.mark.parametrize('data_migration_version', [None])
+@pytest.mark.parametrize('perform_migrations_at_unlock', [False])
+@pytest.mark.parametrize('perform_upgrades_at_unlock', [False])
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+@pytest.mark.parametrize('perform_nodes_insertion', [False])
+def test_migration_6_default_nodes(rotkehlchen_api_server):
```
Comment (yabirgb) [inline rotkehlchen/tests/data_migrations/test_migrations.py:295]:
```suggestion
def test_migration_6_default_rpc_nodes(rotkehlchen_api_server):
```

File: rotkehlchen/tests/data_migrations/test_migrations.py [hunk 7]
```diff
@@ -285,3 +285,97 @@ def test_migration_5(rotkehlchen_api_server):
...
+
+    assert nodes_in_db == new_nodes
+
+
+@pytest.mark.parametrize('data_migration_version', [None])
+@pytest.mark.parametrize('perform_migrations_at_unlock', [False])
+@pytest.mark.parametrize('perform_upgrades_at_unlock', [False])
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+@pytest.mark.parametrize('perform_nodes_insertion', [False])
+def test_migration_6_customized_nodes(rotkehlchen_api_server):
```
Comment (yabirgb) [inline rotkehlchen/tests/data_migrations/test_migrations.py:342]:
```suggestion
def test_migration_6_customized_rpc_nodes(rotkehlchen_api_server):
```

File: rotkehlchen/tests/data_migrations/test_migrations.py [hunk 8]
```diff
@@ -285,3 +285,97 @@ def test_migration_5(rotkehlchen_api_server):
...
+    # Now nodes should be the new defaults
+    with database.conn.read_ctx() as cursor:
+        nodes_in_db = set(cursor.execute('SELECT name, endpoint, owned, weight, blockchain FROM web3_nodes'))  # noqa: E501
+
+    new_nodes = {
+        (node['name'], node['endpoint'], node['owned'], node['weight'], node['blockchain'])  # noqa: E501
+        for node in new_nodes_info
+    }
+
+    assert nodes_in_db == new_nodes
```
Comment (yabirgb) [inline rotkehlchen/tests/data_migrations/test_migrations.py:334]: can you also add a test that verfies that it holds that `sum(weights) == 1` for ethereum nodes?

File: docs/changelog.rst [hunk 9]
```diff
@@ -15,7 +15,9 @@ Changelog
 * :bug:`-` Fix an edge-case problem with synchronizing database with the server.
 * :bug:`-` Users will be able to edit EVM tokens again and the information about underlying tokens will be correctly displayed.
 * :bug:`-` Fix an issue with nginx websocket proxying on docker.
-* :bug:`-` Now `Card Cashback Reversal` entries from cryptocom csv will be imported correctly.
+* :bug:`-` Now `Card Cashback Reversal` entries from cryptocom csv will be correctly imported.
+* :bug:`-` Make ethereum queries faster by removing dead open nodes. If the user has not customized the nodes they will be automatically replaced with the new ones. Otherwise a manual action is needed.
```
Comment (yabirgb) [inline docs/changelog.rst:19]: For users that won't see this list updated maybe we should add something in the release notes saying what are the new rpc endpoints?

```suggestion
* :bug:`-` Users will see the list of ethereum RPC nodes updated if they didn't modify it. For users that already modified this list action is required.
```

File: rotkehlchen/data_migrations/migrations/migration_6.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+            write_cursor=write_cursor,
+            proportion_to_share=ONE,
+            exclude_identifier=None,
+            blockchain=SupportedBlockchain.ETHEREUM,
+        )
+
+
+def data_migration_6(write_cursor: 'DBCursor', rotki: 'Rotkehlchen') -> None:
+    """
+    - Update ethereum nodes nodes in 1.26.2
```
Comment (yabirgb) [inline rotkehlchen/data_migrations/migrations/migration_6.py:75]:
```suggestion
    - Update ethereum rpc nodes in 1.26.2
```

File: rotkehlchen/tests/data_migrations/test_migrations.py [hunk 11]
```diff
@@ -285,3 +285,97 @@ def test_migration_5(rotkehlchen_api_server):
...
+    database = rotki.data.db
+    dir_path = Path(__file__).resolve().parent.parent.parent
+    with open(dir_path / 'data' / 'nodes_as_of_1-26-1.json', 'r') as f:
+        old_nodes_info = json.loads(f.read())
+
+    old_nodes = [
+        [node['name'], node['endpoint'], node['owned'], node['weight'], node['active'], node['blockchain']]  # noqa: E501
+        for node in old_nodes_info
+    ]
+    old_nodes[0][0] = 'Renamed etherscan'  # customize a node
```
Comment (yabirgb) [inline rotkehlchen/tests/data_migrations/test_migrations.py:356]: This node can't be customixed irc. Can you add a new one in the database, for example from the list of new ones and see that it doesn't get deleted?

**Review by yabirgb — COMMENTED** [review]
Some comments

File: rotkehlchen/data/nodes.json (hunk 1 above)
Comment (nebolax) [reply rotkehlchen/data/nodes.json:37]: By "Load time" criteria on ethereumnodes.com
image

File: docs/changelog.rst (hunk 9 above)
Comment (nebolax) [reply docs/changelog.rst:19]: I am not sure why is your suggestion better? Imo it says the same and excludes "make ethereum queries faster"

File: rotkehlchen/tests/data_migrations/test_migrations.py (hunk 11 above)
Comment (nebolax) [reply rotkehlchen/tests/data_migrations/test_migrations.py:356]: Adding another node is a bit more tricky since in requires to reweigh the nodes. I changed this line so that now the cloudflare node is being renamed. Is it okay in your opinion?

File: docs/changelog.rst (hunk 9 above)
Comment (yabirgb) [reply docs/changelog.rst:19]: I was thinking that speed improvements ( I haven't measured) are only a side effect. The real change is updating the list of nodes.

The proposed changes focuses the attention on what was made and identifies the users affected in the same sentence. Then it inform the rest of users not affected that they will have to take action but `a manual action` is wrong afaik.

But let @LefterisJP hope and agree is fine as it is/propose a better wording here

File: rotkehlchen/tests/data_migrations/test_migrations.py (hunk 11 above)
Comment (yabirgb) [reply rotkehlchen/tests/data_migrations/test_migrations.py]: For the intention of testing is required to rebalance? The weights adding one is needed when we call random from python as is a constraint in the library. There is no other place in the logic where that constraint is needed afaik. Also is just calling one extra function if you do it.

I would suggest to add a new node since it will test a more realistic scenario of users having new nodes vs just editing the name (I would do both, editing name and new node)

File: rotkehlchen/data_migrations/migrations/migration_6.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    'https://mainnet-nethermind.blockscout.com/',  # blockscout
+    'https://mainnet.eth.cloud.ava.do/',  # avado
+    'https://web3.1inch.exchange',  # 1inch
+    'https://names.mewapi.io/rpc/eth',  # myetherwallet
+]
+
+
+def update_nodes_in_database(write_cursor: 'DBCursor', rotki: 'Rotkehlchen') -> None:
+    """
+    Updates the ethereum nodes in the database. If nodes in the db are exactly same as the defaults
```
Comment (yabirgb) [inline rotkehlchen/data_migrations/migrations/migration_6.py:29]:
```suggestion
    Updates the ethereum nodes in the database. If nodes in the db are exactly the same as the defaults
```

or `identical`

File: rotkehlchen/tests/data_migrations/test_migrations.py (hunk 11 above)
Comment (nebolax) [reply rotkehlchen/tests/data_migrations/test_migrations.py]: Ok, added also a new node

File: docs/changelog.rst (hunk 9 above)
Comment (LefterisJP) [reply docs/changelog.rst]: In general agree with Yabir. Changelog should be informative for what happened and what should the user do if anything. You are not making all ethereum queries faster withthis change, so dont' make such claims in the changelog.

Change this one to:
> Default list of open nodes has been modified. Users who used only defaults may notice an increase in speed. For others it depends on their selection and they are encouraged to review and modify their ethereum nodes list.

File: rotkehlchen/data_migrations/migrations/migration_6.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    with open(dir_path / 'data' / 'nodes.json', 'r') as f:
+        new_nodes_info = json.loads(f.read())
+
+    old_nodes = {
+        (node['name'], node['endpoint'], node['owned'], node['weight'])
+        for node in old_nodes_info
+    }
+
+    nodes_in_db = set(write_cursor.execute(
+        'SELECT name, endpoint, owned, weight FROM web3_nodes WHERE blockchain=?',
```
Comment (LefterisJP) [inline rotkehlchen/data_migrations/migrations/migration_6.py:45]: perhaps you can also ignore the owned node. This way you get those who also just kept defaults for all open nodes but also added their own node.

File: rotkehlchen/data_migrations/migrations/migration_6.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+        (SupportedBlockchain.ETHEREUM.value,),
+    ))
+
+    if nodes_in_db == old_nodes:
+        # If nodes in the db are exactly same as old defaults, replace them with new defaults
+        log.info('Completely replacing ethereum rpc nodes with new defaults')
+        write_cursor.execute(
+            'DELETE FROM web3_nodes WHERE blockchain=?',
```
Comment (LefterisJP) [inline rotkehlchen/data_migrations/migrations/migration_6.py:53]: if you ignore owned node up there you would also need to do so here too. I would also add an owned node in the tests to see it happens correctly. Perhaps one test with owned node and one without

File: rotkehlchen/data_migrations/migrations/migration_6.py [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+        write_cursor.executemany(
+            'INSERT INTO web3_nodes(name, endpoint, owned, active, weight, blockchain) VALUES (?, ?, ?, ?, ?, ?)',  # noqa: E501
+            new_nodes_tuples,
+        )
+    else:
+        # Else just delete dead nodes
+        log.info('Deleting dead ethereum rpc nodes but keeping the other ones')
+        write_cursor.execute(
+            f'DELETE FROM web3_nodes WHERE endpoint IN ({",".join("?"*len(DEAD_NODES_RPCS))}) AND blockchain=?',  # noqa: E501
+            DEAD_NODES_RPCS + [SupportedBlockchain.ETHEREUM.value],
```
Comment (LefterisJP) [inline rotkehlchen/data_migrations/migrations/migration_6.py:69]: no need to do list adddition here. This is better and gives it as a tuple and is easier to match to the actual arguments
```suggestion
            (*DEAD_NODES_RPCS, SupportedBlockchain.ETHEREUM.value),
```

File: rotkehlchen/data_migrations/migrations/migration_6.py [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+            (node['name'], node['endpoint'], node['owned'], node['active'], str(FVal(node['weight'])), node['blockchain'])  # noqa: E501
+            for node in new_nodes_info
+        ]
+        write_cursor.executemany(
+            'INSERT INTO web3_nodes(name, endpoint, owned, active, weight, blockchain) VALUES (?, ?, ?, ?, ?, ?)',  # noqa: E501
+            new_nodes_tuples,
+        )
+    else:
+        # Else just delete dead nodes
+        log.info('Deleting dead ethereum rpc nodes but keeping the other ones')
```
Comment (LefterisJP) [inline rotkehlchen/data_migrations/migrations/migration_6.py:66]: Change all these info logs to debug, same as what we do in the db upgrades

File: rotkehlchen/data_migrations/migrations/migration_6.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+            exclude_identifier=None,
+            blockchain=SupportedBlockchain.ETHEREUM,
+        )
+
+
+def data_migration_6(write_cursor: 'DBCursor', rotki: 'Rotkehlchen') -> None:
+    """
+    - Update ethereum rpc nodes nodes in 1.26.2
+    """
+    update_nodes_in_database(write_cursor=write_cursor, rotki=rotki)
```
Comment (LefterisJP) [inline rotkehlchen/data_migrations/migrations/migration_6.py:84]: add a start and finish debug log like we do in the upgrades

File: rotkehlchen/tests/data_migrations/test_migrations.py [hunk 18]
```diff
@@ -285,3 +285,101 @@ def test_migration_5(rotkehlchen_api_server):
...
+    dir_path = Path(__file__).resolve().parent.parent.parent
+    with open(dir_path / 'data' / 'nodes_as_of_1-26-1.json', 'r') as f:
+        old_nodes_info = json.loads(f.read())
+    with open(dir_path / 'data' / 'nodes.json', 'r') as f:
+        new_nodes_info = json.loads(f.read())
+
+    old_nodes = [
+        (node['name'], node['endpoint'], node['owned'], node['weight'], node['active'], node['blockchain'])  # noqa: E501
+        for node in old_nodes_info
+    ]
```
Comment (LefterisJP) [inline rotkehlchen/tests/data_migrations/test_migrations.py:310]: All this code repeats between this and the other test. Abstract it out

File: rotkehlchen/tests/data_migrations/test_migrations.py [hunk 19]
```diff
@@ -285,3 +285,101 @@ def test_migration_5(rotkehlchen_api_server):
...
+
+
+@pytest.mark.parametrize('data_migration_version', [None])
+@pytest.mark.parametrize('perform_migrations_at_unlock', [False])
+@pytest.mark.parametrize('perform_upgrades_at_unlock', [False])
+@pytest.mark.parametrize('use_clean_caching_directory', [True])
+@pytest.mark.parametrize('perform_nodes_insertion', [False])
+def test_migration_6_customized_rpc_nodes(rotkehlchen_api_server):
+    """
+    Test that the sixth data migration works when the user has some cusotmized the rpc nodes.
```
Comment (LefterisJP) [inline rotkehlchen/tests/data_migrations/test_migrations.py:345]:
```suggestion
    Test that the sixth data migration works when the user has some customized the rpc nodes.
```

File: rotkehlchen/tests/data_migrations/test_migrations.py [hunk 20]
```diff
@@ -285,3 +285,101 @@ def test_migration_5(rotkehlchen_api_server):
...
+            'INSERT INTO web3_nodes (name, endpoint, owned, weight, active, blockchain) VALUES (?, ?, ?, ?, ?, ?)',  # noqa: E501
+            old_nodes,
+        )
+
+    migration_patch = patch(
+        'rotkehlchen.data_migrations.manager.MIGRATION_LIST',
+        new=[MIGRATION_LIST[5]],
+    )
+    with migration_patch:
+        DataMigrationManager(rotki).maybe_migrate_data()
```
Comment (LefterisJP) [inline rotkehlchen/tests/data_migrations/test_migrations.py:323]: as mentioned in the code part, perhaps add a version where we also have all open nodes as default except for having also an own node.

**Review by LefterisJP — COMMENTED** [review]
Some comments.

**Review by LefterisJP — COMMENTED** [review]
some smaller comments and questions

File: rotkehlchen/tests/data_migrations/test_migrations.py [hunk 21]
```diff
@@ -220,8 +220,8 @@ def test_migration_4(rotkehlchen_api_server):
                     assert web3_node.weight == FVal(node['weight'])
                     continue
         assert len(web3_nodes) >= 5
-        assert web3_nodes[5].node_info.owned is True
-        assert web3_nodes[5].node_info.endpoint == 'https://localhost:5222'
+        assert web3_nodes[4].node_info.owned is True
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/tests/data_migrations/test_migrations.py:223]: Because the default nodes that we have are actually loaded via the 4th migration and when queried from the DB are sorted in alphabetical order by name. Since I changed the nodes, the test also needed an adjustment

File: rotkehlchen/tests/exchanges/test_gemini.py [hunk 22]
```diff
@@ -1,3 +1,4 @@
+# type: ignore
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/tests/exchanges/test_gemini.py:1]: Hmm, yes, no need indeed. I was afraid that mypy will complain about "Unreachable code" for every test function in this file

[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/tests/exchanges/test_gemini.py]: This is now how you skip tests in the file. And you only need to skip the tests that do remote calls. I am going to amend that commit and push.

**Review by LefterisJP — APPROVED** [review]
LGTM.

Only failing test was an NFT one due to opensea timeout.

```[09/12/2022 14:48:16 UTC] DEBUG rotkehlchen.api.server Greenlet-288: end rotki api GET /api/1/nfts view_args={}, query_string=b'', status_code=502, result={'result': None, 'message': 'Reached opensea backoff limit after we incrementally backed off for https://api.opensea.io/api/v1/collection/zklink-by-project-galaxy/stats?name=zklink-by-project-galaxy'}```
